src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
     return std::make_pair(pCache, nCache);
 }
 
-std::uint64_t
-DatabaseShardImp::available() const
+bool
+DatabaseShardImp::sufficientStorage(
+    std::uint32_t numShards,
+    bool isHistorical,
+    std::lock_guard<std::mutex> const&) const
 {
     try
     {
-        return boost::filesystem::space(dir_).available;
+        std::vector<std::uint64_t> capacities;
+
+        // Get the available storage for each historical path
+        if (isHistorical && !historicalPaths_.empty())
+        {
+            std::for_each(
+                historicalPaths_.begin(),
+                historicalPaths_.end(),
+                [&capacities](boost::filesystem::path const& path) {
+                    capacities.push_back(
+                        boost::filesystem::space(path).available);
+                });
+        }
+
+        // Get the available storage for the main shard path
+        else
+            capacities.push_back(boost::filesystem::space(dir_).available);
+
+        return std::find_if(
+                   capacities.begin(),
+                   capacities.end(),
+
+                   // Leverage all the historical shard paths to
+                   // see if collectively they can fit the specified
+                   // number of shards. For this to work properly,
+                   // each historical path must correspond to a separate
+                   // physical device.
+                   [&numShards, this](std::uint64_t const& capacity) {
+                       numShards -= std::min(
+                           capacity / avgShardFileSz_,
+                           std::uint64_t(numShards));
+
+                       return numShards == 0;
+                   }) != capacities.end();
```

Comment:
While clever, we aren't using `find_if` to find anything. It's likely people will need to decode this. What do you think about calculating the available number of shards and checking if there's enough space that way instead. I.e.
```

        std::uint64_t availSpacsShards = 0;
        for (auto const c : capacities)
        {
            availSpacsShards += c / avgShardFileSz_;
        }
        return availSpacsShards >= numShards;
```
That code does not exit early (tho it's easy enough to do that). But given the size of `capacities` it won't matter.

Reply:
Changed to a for loop.